crates/sqlbuiltins/src/functions/scalars/similarity.rs

```diff
@@ -0,0 +1,309 @@
+                    dtype
+                )))
+            }
+        },
+        dtype => {
+            return Err(DataFusionError::Execution(format!(
+                "Unsupported data type for cosine_similarity: {:?}",
+                dtype
+            )))
+        }
+    })
+}
+
+
+impl BuiltinScalarUDF for CosineSimilarity {
+    fn try_as_expr(
+        &self,
+        _: &catalog::session_catalog::SessionCatalog,
+        args: Vec<Expr>,
+    ) -> datafusion::error::Result<Expr> {
+        let scalar_f: ScalarFunctionImplementation = Arc::new(move |args| {
+            let query_vec = match &args[0] {
+                ColumnarValue::Array(arr) => arr_to_query_vec(arr),
+                ColumnarValue::Scalar(ScalarValue::List(arr)) => arr_to_query_vec(arr.as_ref()),
+                ColumnarValue::Scalar(ScalarValue::FixedSizeList(arr)) => {
+                    arr_to_query_vec(arr.as_ref())
+                }
+                _ => {
+                    return Err(DataFusionError::Execution(
+                        "Invalid argument type".to_string(),
+                    ))
+                }
+            }?;
+            let target_vec = match &args[1] {
+                ColumnarValue::Array(arr) => arr_to_target_vec(arr),
+                ColumnarValue::Scalar(ScalarValue::List(arr)) => arr_to_target_vec(arr.as_ref()),
+                ColumnarValue::Scalar(ScalarValue::FixedSizeList(arr)) => {
+                    arr_to_target_vec(arr.as_ref())
+                }
+                _ => {
+                    return Err(DataFusionError::Execution(
+                        "Invalid argument type".to_string(),
+                    ))
+                }
+            }?;
+
+            let dimension = target_vec.value_length() as usize;
+            if query_vec.len() != dimension {
+                return Err(DataFusionError::Execution(
+                    "Query vector and target vector must have the same length".to_string(),
+                ));
+            }
+
+            let result: Arc<dyn Array> = lance_linalg::distance::cosine_distance_arrow_batch(
+                query_vec.as_ref(),
+                &target_vec,
+            )
+            .map_err(|e| DataFusionError::Execution(e.to_string()))?;
+
+            Ok(ColumnarValue::Array(result))
+        });
+
+        let return_type_fn: ReturnTypeFunction = Arc::new(move |_| {
+            let dtype = DataType::Float32;
+            Ok(Arc::new(dtype))
+        });
+
+        let udf = ScalarUDF::new(
+            Self::NAME,
+            &self.signature().unwrap(),
+            &return_type_fn,
+            &scalar_f,
+        );
+
+
+        Ok(Expr::ScalarFunction(ScalarFunction::new_udf(
+            Arc::new(udf),
+            args,
+        )))
+    }
+}
+
+// modified/copied from arrow_cast
+fn cast_list_to_fixed_size_list<OffsetSize>(
+    array: &GenericListArray<OffsetSize>,
+    field: &FieldRef,
+    size: i32,
```

Comment:
why not usize?

Reply:
fixedsizelist datatype expects i32. Using usize would require unnecessary casts inside this function. 
This is just copy/pasted from arrow-cast with the dyn array output replaced with fixedsizearray.